### PR TITLE
[Merged by Bors] - Fix attestation shuffling filter

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Fix a bug in block production that results in blocks with 0 attestations during the first slot of an epoch.

The bug is marked by debug logs of the form:

> DEBG Discarding attestation because of missing ancestor, block_root: 0x3cc00d9c9e0883b2d0db8606278f2b8423d4902f9a1ee619258b5b60590e64f8, pivot_slot: 4042591

It occurs when trying to look up the shuffling decision root for an attestation from a slot which is prior to fork choice's finalized block. This happens frequently when proposing in the first slot of the epoch where we have:

- `current_epoch == n`
- `attestation.data.target.epoch == n - 1`
- attestation shuffling epoch `== n - 3` (decision block being the last block of `n - 3`)
- `state.finalized_checkpoint.epoch == n - 2` (first block of `n - 2` is finalized)

Hence the shuffling decision slot is out of range of the fork choice backwards iterator _by a single slot_.

Unfortunately this bug was hidden when we weren't pruning fork choice, and then reintroduced in v2.5.1 when we fixed the pruning (https://github.com/sigp/lighthouse/releases/tag/v2.5.1). There's no way to turn that off or disable the filtering in our current release, so we need a new release to fix this issue.

Fortunately, it also does not occur on every epoch boundary because of the gradual pruning of fork choice every 256 blocks (~8 epochs):

https://github.com/sigp/lighthouse/blob/01e84b71f524968f5b940fbd2fa31d29408b6581/consensus/proto_array/src/proto_array_fork_choice.rs#L16

https://github.com/sigp/lighthouse/blob/01e84b71f524968f5b940fbd2fa31d29408b6581/consensus/proto_array/src/proto_array.rs#L713-L716

So the probability of proposing a 0-attestation block given a proposal assignment is approximately `1/32 * 1/8 = 0.39%`.

## Proposed Changes

- Load the block's shuffling ID from fork choice and verify it against the expected shuffling ID of the head state. This code was initially written before we had settled on a representation of shuffling IDs, so I think it's a nice simplification to make use of them here rather than more ad-hoc logic that fundamentally does the same thing.

## Additional Info

Thanks to @moshe-blox for noticing this issue and bringing it to our attention.

### CODE REVIEW

**Comment by michaelsproul** [comment]
Before this is ready to merge I would like to:

- Run it on blockdreamer and confirm the absence of 0-attestation blocks (currently I have one every ~8 epochs)
- Fuzz it with Hydra
- Test it manually with some more integration tests

**Comment by michaelsproul** [comment]
Marking this tentatively ready for review. Tests are updated, blockdreamer seems happy, I've just spun up Hydra with 240 workers. I might need to make some Hydra tweaks to check for empty blocks, but will try that next week.

I threw out my approach from yesterday because I realised that the `beacon_block_root` of an attestation isn't necessarily descended from its _source_, so all that reasoning was bunk. I think the new implementation is cleaner and easier to understand anyway.

**Comment by michaelsproul** [comment]
So far so good with Hydra & blockdreamer. I think this is ready to merge once reviewed.

**Review by paulhauner — APPROVED** [review]
I spent quite a while reviewing this and I'm confident that it's correct! Nice work :rocket:

**Comment by michaelsproul** [comment]
Another way of looking at the likelihood of this bug is that ~1/700 mainnet blocks are affected, when all of the following conditions are met:

- The block must be at the first slot of the epoch (1/32 chance).
- The block must be proposed by a Lighthouse validator (~37% chance per [clientdiversity.org](https://clientdiversity.org/)).
- The fork choice data structure must have pruned the epoch prior to finalization (~1/8 chance).

**Pull request closed without merging**
